keygen/src/ssl.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2018 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/pem.h>
+#include <openssl/ec.h>
+#include <openssl/x509.h>
+#include <sodium.h>
+#include <errno.h>
+
+#include "ssl.h"
+#include "enftun/config.h"
+
+/**
+ * public_key_encode_x962() - Encode a public key from an EVP_PKEY structure
+ *
+ * Caller must free the returned buffer
+ *
+ * Return: A new buffer containing the x9.62-encoded string
+ */
+static char *public_key_encode_x962(EVP_PKEY *pkey)
+{
+    EC_KEY *key = NULL;
+    const EC_POINT * point = NULL;
+    char *hex = NULL;
+
+    key = EVP_PKEY_get1_EC_KEY(pkey);
+    if (!key) {
+        printf("%s Could not get EC_KEY from pkey", __func__);
+        goto out;
+    }
+
+    point = EC_KEY_get0_public_key(key);
+    if (!point) {
+        printf("%s Could not get point from EC_KEY", __func__);
+        goto out;
+    }
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
+
+out:
+    EC_KEY_free(key);
+    return hex;
+}
+
+/**
+ * gen_EC_PKEY() - Generate an Elliptic Curve key (NID_X9_62_prime256v1)
+ *
+ * Return: A pointer to a new EC_KEY instance or NULL on failure
+ */
+int gen_key(struct key *key_out)
+{
+    int ret;
+    EC_KEY * ec_key = NULL;
+
+    ec_key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+    if (!ec_key) {
+        fprintf(stderr, "%s ERR: Could not generate EC_KEY", __func__);
+        ret = 0;
+        goto out;
```

Comment:
`ret` will be returned unset here. Maybe just initialize `ret` to `0`?

Reply:
Ret is set to 0 here, nothing wrong with initializing it though.

Reply:
Ha, that's weird, sorry, don't know what I was looking at when I wrote that comment